sail/src/main/java/org/apache/rya/rdftriplestore/inference/InferenceEngine.java

```diff
@@ -363,13 +367,173 @@ public void refreshGraph() throws InferenceEngineException {
             	}
             }
 
+            refreshDomainRange();
+
             refreshPropertyRestrictions();
 
         } catch (QueryEvaluationException e) {
             throw new InferenceEngineException(e);
         }
     }
 
+    /**
+     * Queries domain and range information, then populates the inference engine with direct
+     * domain/range relations and any that can be inferred from the subclass graph, subproperty
+     * graph, and inverse property map. Should be called after that class and property information
+     * has been refreshed.
+     *
+     * Computes indirect domain/range:
+     *  - If p1 has domain c, and p2 is a subproperty of p1, then p2 also has domain c.
+     *  - If p1 has range c, and p2 is a subproperty of p1, then p2 also has range c.
+     *  - If p1 has domain c, and p2 is the inverse of p1, then p2 has range c.
+     *  - If p1 has range c, and p2 is the inverse of p1, then p2 has domain c.
+     *  - If p has domain c1, and c1 is a subclass of c2, then p also has domain c2.
+     *  - If p has range c1, and c1 is a subclass of c2, then p also has range c2.
+     * @throws QueryEvaluationException
+     */
+    private void refreshDomainRange() throws QueryEvaluationException {
+        Map<URI, Set<URI>> domainByTypePartial = new ConcurrentHashMap<>();
+        Map<URI, Set<URI>> rangeByTypePartial = new ConcurrentHashMap<>();
+        // First, populate domain and range based on direct domain/range triples.
+        CloseableIteration<Statement, QueryEvaluationException> iter = RyaDAOHelper.query(ryaDAO, null, RDFS.DOMAIN, null, conf);
+        try {
+            while (iter.hasNext()) {
```

Comment:
It seems like this map building loop is extremely prevalent in the refresh methods for all of the property restrictions.  Is there any way to pull this logic out into a buildRestrictionMap(...) method that takes in the map that is being updated/created/refreshed, and/or the iterator/property URI used to create the iterator?